### PR TITLE
[Data Object] Automatically create redirects for changed URL slugs

### LINKED ISSUE #5767: [Data Object] Automatically create redirects for changed URL slugs
Related to #5766 which has to be done first.

### PR DESCRIPTION
**Pull request by brusch**
Resolves #5767

### CODE REVIEW

File: models/DataObject/ClassDefinition/Data/UrlSlug.php [hunk 1]
```diff
@@ -286,6 +290,42 @@ public function save($object, $params = [])
                 }
             }
         }
+
+        // check for previous slugs and create redirects
+        if(is_array($data)) {
```
Comment (dpfaffenbauer) [inline models/DataObject/ClassDefinition/Data/UrlSlug.php]:
```
if(!is_array($data)) {
     return;
}
```

File: models/DataObject/ClassDefinition/Data/UrlSlug.php [hunk 2]
```diff
@@ -286,6 +290,42 @@ public function save($object, $params = [])
...
+            foreach($data as $slug) {
+                if($previousSlug = $slug->getPreviousSlug()) {
```
Comment (dpfaffenbauer) [inline models/DataObject/ClassDefinition/Data/UrlSlug.php]:
```
if($previousSlug !== $slug->getPreviousSlug()) {
    continue;
}
```

Comment (brusch) [reply models/DataObject/ClassDefinition/Data/UrlSlug.php]: I guess you mean `if($previousSlug === $slug->getSlug()) {` right?

File: models/DataObject/Data/UrlSlug.php [hunk 3]
```diff
@@ -132,11 +137,33 @@ public function getSlug(): ?string
      */
     public function setSlug(?string $slug)
     {
+        if($this->slug && !$this->previousSlug && $slug !== $this->slug) {
+            $this->previousSlug = $this->slug;
```
Comment (weisswurstkanone) [inline models/DataObject/Data/UrlSlug.php]: wouldn't it be better to do this in createFromDataRow?

Comment (brusch) [reply models/DataObject/Data/UrlSlug.php]: But there it's not covering API updates ... so I guess this is the best place, right?

Comment (weisswurstkanone) [reply models/DataObject/Data/UrlSlug.php]: Would still say that previousSlug should only be set when loaded from resource. It should be kind of a dirty detection, which causes a redirect, right ?

Comment (brusch) [reply models/DataObject/Data/UrlSlug.php]: Yep, now I got your point 😉

**Comment by brusch** [comment]
@dvesh3 regarding the need of calling `saveToSession()` on publish: 

In my opinion in would be sufficient to call `Service::removeElementFromSession()` here, since the latest content is then anyway stored in the published document: 
https://github.com/pimcore/pimcore/blob/9b0a4ebdc831dbd5cc21d314388ee34b8d4f8703/bundles/AdminBundle/Controller/Admin/Document/PageController.php#L171

But we can leave it as it is for now, as it doesn't hurt and the session is anyway cleaned up as soon as the document gets closed.

**Pull request merged**